src/EditorFeatures/CSharpTest/MisplacedUsingDirectives/MisplacedUsingDirectivesInCompilationUnitCodeFixProviderTests.cs

```diff
@@ -0,0 +1,418 @@
+
+namespace TestNamespace
+{
+}
+";
+
+            return TestDiagnosticMissingAsync(testCode, InsidePreferPreservationOption);
+        }
+
+        /// <summary>
+        /// Verifies that having using statements in the compilation unit will not produce any diagnostics, nor will
+        /// having using statements inside a namespace.
+        /// </summary>
+        [Fact]
+        public Task WhenPreserve_UsingsInCompilationUnitAndNamespace_ValidUsingStatements()
+        {
+            var testCode = @"[|using System;|]
+
+namespace TestNamespace
+{
+    using System.Threading;
+}
+";
+
+            return TestDiagnosticMissingAsync(testCode, InsidePreferPreservationOption);
+        }
+
+        #endregion
+
+        #region Test InsideNamespace
+
+        /// <summary>
+        /// Verifies that valid using statements in a namespace does not produce any diagnostics.
+        /// </summary>
+        [Fact]
+        public Task WhenInsidePreferred_UsingsInNamespace_ValidUsingStatements()
+        {
+            var testCode = @"namespace TestNamespace
+{
+    [|using System;
+    using System.Threading;|]
+}
+";
+
+            return TestDiagnosticMissingAsync(testCode, InsideNamespaceOption);
+        }
+
+        /// <summary>
+        /// Verifies that having using statements in the compilation unit will not produce any diagnostics when there are type definition present.
+        /// </summary>
+        /// <param name="typeDefinition">The type definition to test.</param>
+        [Theory]
+        [InlineData(ClassDefinition)]
+        [InlineData(StructDefinition)]
+        [InlineData(InterfaceDefinition)]
+        [InlineData(EnumDefinition)]
+        [InlineData(DelegateDefinition)]
+        public Task WhenInsidePreferred_UsingsInCompilationUnitWithTypeDefinition_ValidUsingStatements(string typeDefinition)
+        {
+            var testCode = $@"[|using System;|]
+
+{typeDefinition}
+";
+
+            return TestDiagnosticMissingAsync(testCode, InsideNamespaceOption);
+        }
+
+        /// <summary>
+        /// Verifies that having using statements in the compilation unit will not produce any diagnostics when there are attributes present.
+        /// </summary>
+        [Fact]
+        public Task WhenInsidePreferred_UsingsInCompilationUnitWithAttributes_ValidUsingStatements()
+        {
+            var testCode = @"[|using System.Reflection;|]
+
+[assembly: AssemblyVersion(""1.0.0.0"")]
+
+namespace TestNamespace
+{
+    using System;
+    using System.Threading;
+}
+";
+
+            return TestDiagnosticMissingAsync(testCode, InsideNamespaceOption);
+        }
+
+        /// <summary>
+        /// Verifies that the code fix will move the using directives and not place System directives first.
+        /// </summary>
+        [Fact]
+        public Task WhenInsidePreferred_UsingsInCompilationUnit_UsingsMovedAndSystemPlacedFirstIgnored()
+        {
+            var testCode = @"[|using Microsoft.CodeAnalysis;
```

Comment:
Ah, so we have one diagnostic per misplaced using, but the code fix offers fixing all of them at once, not one at a time. Batch fixer should work fine for this model as FixAll cannot have merge conflicts. Clears thing up for me, thanks.